extensions/amp-content-viewer/0.1/amp-content-viewer.js

```diff
@@ -0,0 +1,671 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as st from '../../../src/style';
+import * as tr from '../../../src/transition';
+import {Animation} from '../../../src/animation';
+import {CSS} from '../../../build/amp-content-viewer-0.1.css';
+import {
+  DoubletapRecognizer,
+  PinchRecognizer,
+  SwipeXYRecognizer,
+  TapRecognizer,
+  TapzoomRecognizer,
+} from '../../../src/gesture-recognizers';
+import {Gestures} from '../../../src/gesture';
+import {Layout} from '../../../src/layout';
+import {bezierCurve} from '../../../src/curve';
+import {clamp} from '../../../src/utils/math';
+import {continueMotion} from '../../../src/motion';
+import {createCustomEvent} from '../../../src/event-helper';
+import {dev, user} from '../../../src/log';
+import {isExperimentOn} from '../../../src/experiments';
+import {
+  layoutRectFromDomRect,
+  layoutRectLtwh,
+} from '../../../src/layout-rect';
+
+const PAN_ZOOM_CURVE_ = bezierCurve(0.4, 0, 0.2, 1.4);
+const TAG = 'amp-content-viewer';
+const DEFAULT_MAX_SCALE = 3;
+const MAX_ANIMATION_DURATION = 250;
+
+
+const ELIGIBLE_TAGS = {
+  'SVG': true,
+  'DIV': true,
+  'AMP-IMG': true,
+};
+
+const SUPPORT_VALIDATION_MSG = `amp-content-viewer should
+  have its target element as the one and only child`;
+
+export class AmpContentViewer extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+
+    super(element);
+
+    /** @private {?Element} */
+    this.content_ = null;
+
+    /** @private {number} */
+    this.sourceWidth_ = 0;
+
+    /** @private {number} */
+    this.sourceHeight_ = 0;
+
+    /** @private {?../../../src/layout-rect.LayoutRectDef} */
+    this.elementBox_ = null;
+
+    /** @private {?../../../src/layout-rect.LayoutRectDef} */
+    this.contentBox_ = null;
+
+    /** @private {?UnlistenDef} */
+    this.unlistenOnSwipePan_ = null;
+
+    /** @private */
+    this.scale_ = 1;
```

Comment:
FYI: I am skipping the parts that are shared with existing already reviewed image-viewer.